scripts/machine_translation/dataprocessor.py

```diff
@@ -224,44 +215,72 @@ def make_dataloader(data_train, data_val, data_test, args,
         bucket_scheme = nlp.data.ExpWidthBucket(bucket_len_step=1.2)
     else:
         raise NotImplementedError
-    train_batch_sampler = nlp.data.FixedBucketSampler(lengths=data_train_lengths,
-                                                      batch_size=args.batch_size,
-                                                      num_buckets=args.num_buckets,
-                                                      ratio=args.bucket_ratio,
-                                                      shuffle=True,
-                                                      use_average_length=use_average_length,
-                                                      num_shards=num_shards,
-                                                      bucket_scheme=bucket_scheme)
-    logging.info('Train Batch Sampler:\n%s', train_batch_sampler.stats())
-    train_data_loader = nlp.data.ShardedDataLoader(data_train,
-                                                   batch_sampler=train_batch_sampler,
-                                                   batchify_fn=train_batchify_fn,
-                                                   num_workers=num_workers)
-
-    val_batch_sampler = nlp.data.FixedBucketSampler(lengths=target_val_lengths,
+
+    if dataset_type == 'train':
+        data_lengths = get_data_lengths(data_set)
+        train_batchify_fn = btf.Tuple(btf.Pad(), btf.Pad(),
+                                      btf.Stack(dtype='float32'), btf.Stack(dtype='float32'))
+
+        batch_sampler = nlp.data.FixedBucketSampler(lengths=data_lengths,
+                                                    batch_size=args.batch_size,
+                                                    num_buckets=args.num_buckets,
+                                                    ratio=args.bucket_ratio,
+                                                    shuffle=True,
+                                                    use_average_length=use_average_length,
+                                                    num_shards=num_shards,
+                                                    bucket_scheme=bucket_scheme)
+    else:
+        data_lengths = get_data_lengths(data_set)
+        target_lengths = list(map(lambda x: x[-1], data_lengths))
+        test_batchify_fn = btf.Tuple(btf.Pad(), btf.Pad(),
+                                     btf.Stack(dtype='float32'), btf.Stack(dtype='float32'),
+                                     btf.Stack())
+
+        batch_sampler = nlp.data.FixedBucketSampler(lengths=target_lengths,
```

Comment:
the `nlp.data.FixedBucketSampler` for val and test is very similar to the one for train. The differences only include `shuffle=True` and `batch_size=args.batch_size`. So it is better to merge these two parts to reduce the redundancy.